fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2AnalyticRule.java

```diff
@@ -0,0 +1,869 @@
+                                    .map(ScalarOperator::<ColumnRefOperator>cast)
+                                    .map(columnRefOperator -> new Ordering(columnRefOperator, true, true))
+                                    .collect(Collectors.toList()))
+                    .build();
+        }
+
+        /**
+         * Reform the outer block's predicate into two parts, one contains the subquery conjunct
+         * while the others forms another part.
+         */
+        private void rewriteOuterPredicate() {
+            LogicalFilterOperator filterOperator = filter.getOp().cast();
+            ScalarOperator predicate = filterOperator.getPredicate();
+            List<ScalarOperator> conjuncts = Utils.extractConjuncts(predicate);
+            conjuncts.removeIf(conjunct -> Objects.equals(outerSubqueryConjunct, conjunct));
+            predicateBeforeWindow = Utils.compoundAnd(conjuncts);
+
+            SubqueryReplaceShuttle visitor = new SubqueryReplaceShuttle(applyOp, peerColumnRefMap);
+            predicateAfterWindow = outerSubqueryConjunct.accept(visitor, null);
+        }
+
+        private List<OptExpression> getNewExpressionRoot() {
+            OptExpression beforeWindowFilter =
+                    OptExpression.create(new LogicalFilterOperator(predicateBeforeWindow), apply.inputAt(0));
+            OptExpression window = OptExpression.create(windowOp, beforeWindowFilter);
+            OptExpression afterWindowFilter =
+                    OptExpression.create(new LogicalFilterOperator(predicateAfterWindow), window);
+
+            Map<ColumnRefOperator, ScalarOperator> newProjectMap = Maps.newHashMap(projectOp.getColumnRefMap());
+            windowOp.getWindowCall().keySet()
+                    .forEach(columnRefOperator -> newProjectMap.put(columnRefOperator, columnRefOperator));
+            subAggOp.getAggregations().keySet()
+                    .forEach(newProjectMap::remove);
+            newProjectMap.remove(applyOp.getOutput());
+            OptExpression newProject = OptExpression.create(new LogicalProjectOperator.Builder()
+                    .withOperator(project.getOp().cast())
+                    .setColumnRefMap(newProjectMap)
+                    .build(), afterWindowFilter);
+
+            return Collections.singletonList(newProject);
+        }
+
+        private void collectInLevelOrder(OptExpression root, List<LogicalOperator> collect) {
+            Queue<OptExpression> queue = Lists.newLinkedList();
+            queue.offer(root);
+            while (!queue.isEmpty()) {
+                List<OptExpression> curLevel = Lists.newArrayList(queue);
+                queue.clear();
+                for (OptExpression opt : curLevel) {
+                    collect.add(opt.getOp().cast());
+                    opt.getInputs().forEach(queue::offer);
+                }
+            }
+        }
+
+        private List<Table> getAllTables(List<LogicalOperator> ops) {
```

Comment:
Should we check the table ids pr relation ids?

Reply:
The following sql is not supported right now, which will introduce much complexity.

```sql
select * from t0, t0 as t1
where t0.v1 = t1.v1
and t0.v2 < 5 and t1.v2 > 10
and t0.v3 < (
    select max(t1.v3) from t0 as t1
    where t0.v1 = t1.v1 and t1.v2 > 10
)
```

Besides, comments and tests are added